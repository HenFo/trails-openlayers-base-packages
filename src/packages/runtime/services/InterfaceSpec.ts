export interface InterfaceSpec {
    interfaceName: string;
    qualifier?: string | undefined;
}

export function renderInterfaceSpec({ interfaceName, qualifier }: InterfaceSpec): string {
    return `'${interfaceName}'` + (qualifier ? ` (qualifier: '${qualifier}')` : "");
}
