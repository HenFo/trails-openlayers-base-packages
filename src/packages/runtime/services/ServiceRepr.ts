import { ErrorId } from "./../errors";
import { InterfaceReferenceMetadata, ServiceMetadata } from "../Metadata";
import { Service, ServiceConstructor, ServiceOptions } from "../Service";
import { Error } from "@open-pioneer/core";

export type Dependency = { name: string } & InterfaceReferenceMetadata;

export type ServiceState = "not-constructed" | "constructing" | "constructed" | "destroyed";

/**
 * Represents metadata and state of a service in the runtime.
 * `this.instance` is the actual service instance (when constructed).
 */
export class ServiceRepr {
    static parse(bundleName: string, data: ServiceMetadata): ServiceRepr {
        const clazz = data.clazz;
        const name = data.name;
        const dependencies = Object.entries(data.references ?? {}).map<Dependency>(
            ([name, referenceMetadata]) => {
                return {
                    name,
                    ...referenceMetadata
                };
            }
        );
        // TODO: Properties in metadata?
        const interfaces = (data.provides ?? []).map((p) => p.interface);
        return new ServiceRepr(name, bundleName, clazz, dependencies, interfaces, {});
    }

    /** Unique id of this service. Contains the bundle name and the service name. */
    readonly id: string;

    /** Name of this service in it's bundle. */
    readonly name: string;

    /** Name of the parent bundle. */
    readonly bundleName: string;

    /** Service properties made available via the service's constructor. */
    readonly properties: Readonly<Record<string, unknown>>;

    /** Dependencies required by the service constructor. */
    readonly dependencies: readonly Dependency[];

    /** Interfaces provided by the service. */
    readonly interfaces: readonly string[];

    /** Number of references to this service. */
    private useCount = 0;

    /** Service constructor. */
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    private clazz: ServiceConstructor<any>;

    /** Current state of this service. "constructed" -> instance is available. */
    private _state: ServiceState = "not-constructed";

    /** Service instance, once constructed. */
    private _instance: Service | undefined = undefined;

    constructor(
        name: string,
        bundleName: string,
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        clazz: ServiceConstructor<any>,
        dependencies: Dependency[],
        interfaces: string[],
        properties: Record<string, unknown>
    ) {
        this.id = `${bundleName}::${name}`;
        this.name = name;
        this.bundleName = bundleName;
        this.clazz = clazz;
        this.dependencies = dependencies;
        this.interfaces = interfaces;
        this.properties = properties;
    }

    /** Returns the current service instance or undefined if the service has not been constructed. */
    get instance(): Service | undefined {
        return this._instance;
    }

    /** Returns the current state of the service. */
    get state() {
        return this._state;
    }

    /**
     * Same as `instance`, but throws when the instance has not been constructed.
     */
    getInstanceOrThrow() {
        const instance = this._instance;
        if (!instance) {
            throw new Error(ErrorId.INTERNAL, "Expected service instance to be present.");
        }
        return instance;
    }

    beforeCreate() {
        if (this._state === "not-constructed") {
            this._state = "constructing";
        } else {
            throw new Error(
                ErrorId.INTERNAL,
                "Inconsistent state: service is already under construction."
            );
        }
    }

    /**
     * Instantiates the service by invoking the service constructor 
     * with the given `options`.
     * 
     * The service's use count is initialized to `1`, so every `create()`
     * should be paired with a `removeRef()`.
     * 
     * `destroy()` can be invoked once the final `removeRef()` has returned zero. 
     */
    create(options: ServiceOptions) {
        if (this._state !== "constructing" || this.instance !== undefined) {
            throw new Error(
                ErrorId.INTERNAL,
                "Inconsistent state: service is not being constructed."
            );
        }
        try {
            this._instance = new this.clazz(options);
            this._state = "constructed";
            this.useCount = 1;
            return this._instance;
        } catch (e) {
            throw new Error(
                ErrorId.SERVICE_CONSTRUCTION_FAILED,
                `Failed to create service '${this.id}'.`,
                { cause: e }
            );
        }
    }

    destroy() {
        if (this._instance) {
            try {
                this._instance.destroy?.();
            } catch (e) {
                throw new Error(
                    ErrorId.SERVICE_DESTRUCTION_FAILED,
                    `Failed to destroy service '${this.id}'.`,
                    { cause: e }
                );
            }
        }
        this._instance = undefined;
        this._state = "destroyed";
    }

    /**
     * Adds a use to the service's use count.
     * References to a service are tracked: it should only be destroyed when it is no longer being used.
     */
    addRef() {
        return this.useCount += 1;
    }

    /**
     * Removes a use from the service's use count.
     * Returns the new use count.
     */
    removeRef() {
        return this.useCount -= 1;
    }
}
