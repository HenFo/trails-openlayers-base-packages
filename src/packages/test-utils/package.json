{
    "name": "@open-pioneer/test-utils",
    "private": true,
    "dependencies": {
        "@testing-library/dom": "^8.20.0"
    },
    "peerDependencies": {
        "@open-pioneer/runtime": "workspace:^"
    },
    "devDependencies": {
        "@open-pioneer/runtime": "workspace:^"
    }
}
