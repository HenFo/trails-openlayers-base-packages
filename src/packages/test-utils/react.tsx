import { Service } from "@open-pioneer/runtime";
import {
    PackageContext as InternalPackageContext,
    PackageContextMethods
} from "@open-pioneer/runtime/react-integration";
import { FC, ReactNode, useMemo } from "react";

export type AnyService = Service<Record<string, unknown>>;

export interface PackageContextProviderProps {
    /** Interface implementations, keyed by interface name. */
    services?: {
        [interfaceName: string]: AnyService;
    };

    /** Interface implementations, keyed by interface name and then by qualifier. */
    qualifiedServices?: {
        [interfaceName: string]: {
            [qualifier: string]: AnyService;
        };
    };

    /** Package properties (keyed by package name). */
    properties?: {
        [packageName: string]: Record<string, unknown>;
    };

    /** Children to render */
    children?: ReactNode;
}

/**
 * Injects services and properties into the component tree.
 * React components using pioneer hooks like `useService` and `useProperties`
 * will receive the mocked properties here instead.
 */
export const PackageContextProvider: FC<PackageContextProviderProps> = (props) => {
    const { services, qualifiedServices, properties, children } = props;
    const contextMethods = useMemo(
        () => createPackageContextMethods({ services, qualifiedServices, properties }),
        [services, qualifiedServices, properties]
    );
    return (
        <InternalPackageContext.Provider value={contextMethods}>
            {children}
        </InternalPackageContext.Provider>
    );
};

function createPackageContextMethods(
    options: Omit<PackageContextProviderProps, "children">
): PackageContextMethods {
    const services = options?.services ?? {};
    const qualifiedServices = options?.qualifiedServices ?? {};
    const properties = options?.properties ?? {};
    return {
        getService(packageName, interfaceName, options) {
            if (!options.qualifier) {
                const service = services[interfaceName];
                if (service) {
                    return service;
                }
                const qualified = qualifiedServices[interfaceName];
                if (qualified) {
                    for (const value of Object.values(qualified)) {
                        return value; // Return any implementation
                    }
                }
                throw new Error(
                    `Interface name not bound for testing: '${interfaceName}'. Update the configuration of PackageContextProvider.`
                );
            }

            const qualifiedService = qualifiedServices[interfaceName]?.[options.qualifier];
            if (!qualifiedService) {
                throw new Error(
                    `Interface name not bound for testing: '${interfaceName}' (qualifier '${options.qualifier}'). Update the configuration of PackageContextProvider.`
                );
            }
            return qualifiedService;
        },
        getServices(packageName, interfaceName) {
            const unqualified = services[interfaceName];
            const results = Object.values(qualifiedServices[interfaceName] ?? {});
            if (unqualified) {
                results.push(unqualified);
            }
            return results;
        },
        getProperties(packageName) {
            const packageProperties = properties[packageName];
            if (!packageProperties) {
                throw new Error(
                    `No properties for package '${packageName}' bound for testing. Update the configuration of PackageContextProvider.`
                );
            }
            return packageProperties;
        },
        getIntl(packageName) {
            // TODO
            throw new Error(`No i18n messages for package '${packageName}' bound for testing.`);
        }
    };
}
