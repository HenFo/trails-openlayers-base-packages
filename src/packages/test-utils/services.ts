import { Service, ServiceConstructor } from "@open-pioneer/runtime";

/**
 * Options for the {@link createService} function.
 */
export interface CreateServiceOptions<References> {
    /**
     * References provided to the service constructor.
     *
     * @default {}
     */
    references?: PartialServiceReferences<References>;

    /**
     * Properties provided to the service constructor.
     *
     * @default {}
     */
    properties?: Record<string, unknown>;
}

type PartialServiceReferences<References> = {
    [referenceName in keyof References]?: Partial<References[referenceName]>;
};

/**
 * Creates a new service instance of `clazz`.
 *
 * The options passed into the constructor of `clazz` are taking from the `options` parameter.
 *
 * This function allows the user to define mock- or test objects as service references.
 *
 * Most type-checking features are disabled for convenience, but by default this function's types
 * enforce that a value provided for a reference must match the partial interface of the actually required type.
 *
 * For example, if a service depends on `{ ctx: ApplicationContext }`, this function ensures
 * that the test value for `ctx` implements `Partial<ApplicationContext>`.
 * This can of course be disabled by casting the value explicitly.
 *
 * @param clazz a service class
 * @param options options passed to the service constructor
 * @returns a new instance of the given service class
 */
export async function createService<References extends {}, Interface extends {}>(
    clazz: ServiceConstructor<References, Interface>,
    options?: CreateServiceOptions<References>
): Promise<Service<Interface>> {
    return new clazz({
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        references: (options?.references ?? {}) as any,
        properties: options?.properties ?? {}
    });
}
