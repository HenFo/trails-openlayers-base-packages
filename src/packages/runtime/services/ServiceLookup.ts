import { Error } from "@open-pioneer/core";
import { ErrorId } from "../errors";
import { InterfaceSpec } from "./InterfaceSpec";
import { ServiceRepr } from "./ServiceRepr";

interface Services {
    // The unqualified service implementation (if any).
    unqualified: ServiceRepr | undefined;

    // Services indexed by their qualifier (if any).
    byQualifier: Map<string, ServiceRepr>;
}

export type ReadonlyServiceLookup = Pick<ServiceLookup, "lookup" | "serviceCount">;

export interface Unimplemented {
    type: "unimplemented";
}

export interface Ambiguous {
    type: "ambiguous";
    choices: ServiceRepr[];
}

export interface Found {
    type: "found";
    service: ServiceRepr;
}

export type ServiceLookupResult = Unimplemented | Ambiguous | Found;

export class ServiceLookup {
    // Service implementations indexed by interface name.
    private services = new Map<string, Services>();

    // Total service count
    private _count = 0;

    /**
     * Returns the total number of registered services.
     */
    get serviceCount() {
        return this._count;
    }

    /**
     * Attempts to register the given service as an implementation of `interfaceName`.
     * Throws if the registration did not complete successfully.
     *
     * All services should be registered before any lookups are attempted.
     */
    register(service: ServiceRepr, { interfaceName, qualifier }: InterfaceSpec): void {
        if (!interfaceName) {
            throw new Error(
                ErrorId.INVALID_METADATA,
                `Service '${service.id}' provides invalid interface '${interfaceName}'.`
            );
        }
        if (qualifier != null && !qualifier) {
            throw new Error(
                ErrorId.INVALID_METADATA,
                `Service '${service.id}' uses an invalid qualifier for interface '${interfaceName}': '${qualifier}'.`
            );
        }

        const services = this.ensureInterfaceEntry(interfaceName);
        if (!qualifier) {
            if (services.unqualified) {
                throw new Error(
                    ErrorId.DUPLICATE_INTERFACE,
                    `Cannot register '${service.id}' as interface '${interfaceName}'.` +
                        ` '${interfaceName}' is already provided by service '${services.unqualified.id}'.` +
                        ` If you intend you have multiple implementations in your application, you must use the 'qualifier' attribute.`
                );
            }
            services.unqualified = service;
            ++this._count;
            return;
        }

        const existingQualified = services.byQualifier.get(qualifier);
        if (existingQualified) {
            throw new Error(
                ErrorId.DUPLICATE_INTERFACE,
                `Cannot register '${service.id}' as interface '${interfaceName}' with qualifier '${qualifier}'.` +
                    ` The interface is already provided by service '${existingQualified.id}'.` +
                    ` You can choose a different qualifier or remove the service to resolve the collision.`
            );
        }
        services.byQualifier.set(qualifier, service);
        ++this._count;
    }

    /**
     * Attempts to find a service implementing the given interface.
     */
    lookup({ interfaceName, qualifier }: InterfaceSpec): ServiceLookupResult {
        if (!interfaceName) {
            throw new Error(
                ErrorId.INVALID_METADATA,
                `Invalid interface name during service lookup: '${interfaceName}'.`
            );
        }
        if (qualifier != null && !qualifier) {
            throw new Error(
                ErrorId.INVALID_METADATA,
                `Invalid qualifier during service lookup of interface '${interfaceName}': '${qualifier}'.`
            );
        }

        const services = this.services.get(interfaceName);
        if (!services) {
            return { type: "unimplemented" };
        }

        if (!qualifier) {
            // Just pick one, but the choice must be unambiguous.
            const count = (services.unqualified ? 1 : 0) + services.byQualifier.size;
            if (count > 1) {
                return {
                    type: "ambiguous",
                    choices: (services.unqualified ? [services.unqualified] : []).concat(
                        Array.from(services.byQualifier.values())
                    )
                };
            }

            const service = services.unqualified ?? first(services.byQualifier);
            if (!service) {
                return { type: "unimplemented" };
            }
            return { type: "found", service: service };
        }

        const service = services.byQualifier.get(qualifier);
        if (!service) {
            return { type: "unimplemented" };
        }
        return { type: "found", service: service };
    }

    private ensureInterfaceEntry(interfaceName: string): Services {
        const services = this.services;
        let entry = services.get(interfaceName);
        if (!entry) {
            entry = {
                unqualified: undefined,
                byQualifier: new Map()
            };
            services.set(interfaceName, entry);
        }
        return entry;
    }
}

function first<K, V>(map: Map<K, V>): V | undefined {
    return map.values().next().value;
}
