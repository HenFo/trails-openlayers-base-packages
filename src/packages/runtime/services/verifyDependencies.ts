import { Error } from "@open-pioneer/core";
import { ErrorId } from "../errors";
import { InterfaceSpec, renderInterfaceSpec } from "./InterfaceSpec";
import { ReadonlyServiceLookup, ServiceLookup } from "./ServiceLookup";
import { renderAmbiguousServiceChoices, ServiceRepr } from "./ServiceRepr";

export interface DependencyOptions {
    services?: readonly ServiceRepr[];
    uiDependencies?: readonly UIDependency[];
}

export interface UIDependency extends InterfaceSpec {
    readonly packageName: string;
}

/**
 * Visits all services and ensures that their dependencies can be satisfied by each other without a cycle.
 *
 * Throws an error if a cycle is detected or if a required service is never implemented.
 *
 * @returns the verified index from interface name to service class.
 */
export function verifyDependencies(options: DependencyOptions): ReadonlyServiceLookup {
    const verifier = new Verifier(options);
    verifier.verify();
    return verifier.getServiceIndex();
}

interface GraphItem {
    service: ServiceRepr;
    state: ItemState;
}

type ItemState = "not-visited" | "pending" | "done";

/** The reason why a graph node is being visited. */
type ItemVisitReason =
    | "root"
    | ({
          type: "ui-reference";
          packageName: string;
      } & InterfaceSpec)
    | ({
          type: "service-reference";
          referenceName: string;
      } & InterfaceSpec);

type InterfaceReference =
    | ({
          type: "ui-reference";
          packageName: string;
      } & InterfaceSpec)
    | ({
          type: "service-reference";
          service: ServiceRepr;
          referenceName: string;
      } & InterfaceSpec);

class Verifier {
    private uiDependencies: readonly UIDependency[];
    private items: GraphItem[];
    private serviceLookup = new ServiceLookup();
    private serviceToGraphItem = new Map<ServiceRepr, GraphItem>();

    // Stack of visited nodes
    private stack: [item: GraphItem, reason: ItemVisitReason][] = [];

    constructor(options: DependencyOptions) {
        this.uiDependencies = options.uiDependencies ?? [];
        const services = options.services ?? [];
        const items = (this.items = services.map<GraphItem>((service) => {
            return {
                service: service,
                state: "not-visited"
            };
        }));

        for (const item of items) {
            for (const spec of item.service.interfaces) {
                this.registerService(item, spec);
            }
        }
    }

    verify() {
        for (const uiDependency of this.uiDependencies) {
            this.visitInterface({
                type: "ui-reference",
                ...uiDependency
            });
        }
        for (const item of this.items) {
            this.visitItem(item, "root");
        }
    }

    getServiceIndex() {
        return this.serviceLookup;
    }

    private visitItem(item: GraphItem, reason: ItemVisitReason) {
        const stack = this.stack;
        const state = item.state;
        if (state === "done") {
            return; // Item is initialized, cycle impossible
        }
        if (state === "pending") {
            this.throwCycleError(item, reason);
        }

        stack.push([item, reason]);
        item.state = "pending";
        for (const dep of item.service.dependencies) {
            this.visitInterface({
                type: "service-reference",
                service: item.service,
                ...dep
            });
        }
        item.state = "done";
        stack.pop();
    }

    private visitInterface(ref: InterfaceReference) {
        const childItem = this.findService(ref);

        let reason: ItemVisitReason;
        switch (ref.type) {
            case "service-reference":
                reason = {
                    type: "service-reference",
                    interfaceName: ref.interfaceName,
                    referenceName: ref.referenceName
                };
                break;
            case "ui-reference":
                reason = {
                    type: "ui-reference",
                    interfaceName: ref.interfaceName,
                    packageName: ref.packageName
                };
                break;
        }

        this.visitItem(childItem, reason);
    }

    private throwCycleError(item: GraphItem, reason: ItemVisitReason) {
        // Find the item on the stack (it must be there because we detected a cycle).
        const stack = this.stack;
        const ownIndex = stack.findIndex((entry) => entry[0] === item);
        if (ownIndex === -1) {
            throw new Error(ErrorId.INTERNAL, "Failed to find cycle participant on the stack.");
        }

        const cycleInfo = stack
            .slice(ownIndex)
            .concat([[item, reason]])
            .map<string>((entry, index) => {
                const item = entry[0];
                const reason = index === 0 ? "root" : entry[1]; // info before the cycle does not matter

                let message = `'${item.service.id}'`;
                if (typeof reason === "object") {
                    switch (reason.type) {
                        case "service-reference":
                            message += ` ('${reason.referenceName}' providing '${reason.interfaceName}')`;
                            break;
                        case "ui-reference":
                            message += ` (UI of package '${reason.packageName}' requiring '${reason.interfaceName}')`;
                            break;
                    }
                }
                return message;
            });
        throw new Error(
            ErrorId.DEPENDENCY_CYCLE,
            `Detected dependency cycle: ${cycleInfo.join(" => ")}.`
        );
    }

    private registerService(item: GraphItem, spec: InterfaceSpec) {
        this.serviceLookup.register(item.service, spec);
        this.serviceToGraphItem.set(item.service, item);
    }

    private findService(ref: InterfaceReference) {
        const lookupResult = this.serviceLookup.lookup(ref);
        switch (lookupResult.type) {
            case "unimplemented": {
                const interfaceText = renderInterfaceSpec(ref);
                const message =
                    ref.type === "service-reference"
                        ? `Service '${ref.service.id}' requires an unimplemented interface ${interfaceText} as dependency '${ref.referenceName}'.`
                        : `The UI of package '${ref.packageName}' requires an unimplemented interface ${interfaceText}.`;
                throw new Error(ErrorId.INTERFACE_NOT_FOUND, message);
            }
            case "ambiguous": {
                const interfaceText = renderInterfaceSpec(ref);
                const serviceChoices = renderAmbiguousServiceChoices(lookupResult.choices);
                let message =
                    ref.type === "service-reference"
                        ? `Service '${ref.service.id}' requires the ambiguous interface ${interfaceText} as dependency '${ref.referenceName}'.`
                        : `The UI of package '${ref.packageName}' requires the ambiguous interface ${interfaceText}.`;
                message += ` Possible choices are ${serviceChoices}.`;
                throw new Error(ErrorId.AMBIGUOUS_DEPENDENCY, message);
            }
        }
        const foundService = lookupResult.service;
        const graphItem = this.serviceToGraphItem.get(foundService);
        if (!graphItem || graphItem.service !== foundService) {
            throw new Error(
                ErrorId.INTERNAL,
                `Failed to find matching graph item for service '${foundService.id}'.`
            );
        }
        return graphItem;
    }
}
