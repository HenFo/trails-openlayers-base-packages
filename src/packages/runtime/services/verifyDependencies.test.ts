import { assert, expect, it } from "vitest";
import { expectError } from "../test-utils/expectError";
import { InterfaceSpec } from "./InterfaceSpec";
import { ReadonlyServiceLookup } from "./ServiceLookup";
import { ServiceDependency, ServiceRepr } from "./ServiceRepr";
import { verifyDependencies } from "./verifyDependencies";

it("does not return an error on acyclic graphs", function () {
    const services = mockServices([
        {
            name: "Map",
            package: "map",
            provides: ["services.Map"],
            requires: []
        },
        {
            name: "ExampleTool",
            package: "tools",
            provides: [],
            requires: ["services.Map"]
        }
    ]);

    const lookup = verifyDependencies({ services: services });
    assert.strictEqual(lookup.serviceCount, 1);

    const service = getService(lookup, "services.Map");
    assert.strictEqual(service.id, "map::Map");
});

it("throws when a service is not implemented", function () {
    const services = mockServices([
        {
            name: "ExampleTool",
            package: "tools",
            provides: [],
            requires: ["services.Map"]
        }
    ]);

    const message = expectError(() =>
        verifyDependencies({
            services: services,
            uiDependencies: []
        })
    ).message;
    expect(message).toMatchSnapshot();
});

it("throws when an interface is implemented multiple times", function () {
    const services = mockServices([
        {
            name: "Map1",
            package: "map",
            provides: ["services.Map"],
            requires: []
        },
        {
            name: "Map2",
            package: "map",
            provides: ["services.Map"],
            requires: []
        }
    ]);

    const message = expectError(() =>
        verifyDependencies({
            services: services,
            uiDependencies: []
        })
    ).message;
    expect(message).toMatchSnapshot();
});

it("allows multiple implementations if the services use a 'qualifier' for disambiguation", function () {
    const services = mockServices([
        {
            name: "Map1",
            package: "map",
            provides: [
                {
                    interfaceName: "services.Map",
                    qualifier: "map1"
                }
            ]
        },
        {
            name: "Map2",
            package: "map",
            provides: [
                {
                    interfaceName: "services.Map",
                    qualifier: "map2"
                }
            ]
        }
    ]);

    const lookup = verifyDependencies({
        services: services,
        uiDependencies: []
    });
    assert.strictEqual(lookup.serviceCount, 2);

    const map1 = getService(lookup, "services.Map", "map1");
    assert.strictEqual(map1.id, "map::Map1");

    const map2 = getService(lookup, "services.Map", "map2");
    assert.strictEqual(map2.id, "map::Map2");
});

it("throws for ambiguous service reference", function () {
    const services = mockServices([
        {
            name: "Map1",
            package: "map",
            provides: [
                {
                    interfaceName: "services.Map",
                    qualifier: "map1"
                }
            ]
        },
        {
            name: "Map2",
            package: "map",
            provides: [
                {
                    interfaceName: "services.Map",
                    qualifier: "map2"
                }
            ]
        },
        {
            name: "MapUser",
            package: "map-user",
            requires: [
                {
                    interfaceName: "services.Map"
                }
            ]
        }
    ]);

    const message = expectError(() =>
        verifyDependencies({
            services: services,
            uiDependencies: []
        })
    ).message;
    expect(message).toMatchSnapshot();
});

it("allows to pick an unambiguous implementation via classifier", function () {
    const services = mockServices([
        {
            name: "Map1",
            package: "map",
            provides: [
                {
                    interfaceName: "services.Map",
                    qualifier: "map1"
                }
            ]
        },
        {
            name: "Map2",
            package: "map",
            provides: [
                {
                    interfaceName: "services.Map",
                    qualifier: "map2"
                }
            ]
        },
        {
            name: "MapUser",
            package: "map-user",
            requires: [
                {
                    interfaceName: "services.Map",
                    qualifier: "map2"
                }
            ]
        }
    ]);

    verifyDependencies({
        services: services,
        uiDependencies: []
    });
});

it("throws when a component directly depends on itself", function () {
    const services = mockServices([
        {
            name: "Map",
            package: "map",
            provides: ["services.Map"],
            requires: ["services.Map"]
        }
    ]);
    const message = expectError(() =>
        verifyDependencies({
            services: services
        })
    ).message;
    expect(message).toMatchSnapshot();
});

it("throws when a component depends on itself via a larger cycle", function () {
    const services = mockServices([
        {
            name: "a",
            package: "A",
            provides: ["A"],
            requires: ["B"]
        },
        {
            name: "b",
            package: "B",
            provides: ["B"],
            requires: ["D", "C"]
        },
        {
            name: "c",
            package: "C",
            provides: ["C"],
            requires: ["D", "D", "A"]
        },
        {
            name: "d",
            package: "D",
            provides: ["D"]
        }
    ]);

    const message = expectError(() =>
        verifyDependencies({
            services: services
        })
    ).message;
    expect(message).toMatchSnapshot();
});

it("does not return an error when the UI requires an existing interface", function () {
    const services = mockServices([
        {
            name: "Map",
            package: "map",
            provides: ["services.Map"]
        }
    ]);

    const lookup = verifyDependencies({
        services: services,
        uiDependencies: [
            {
                interfaceName: "services.Map",
                packageName: "foo"
            }
        ]
    });
    assert.strictEqual(lookup.serviceCount, 1);

    const service = getService(lookup, "services.Map");
    assert.strictEqual(service.id, "map::Map");
});

it("throws when the ui requires an interface that is not implemented", function () {
    const services = mockServices([
        {
            name: "Map",
            package: "map",
            provides: ["services.Map"],
            requires: []
        }
    ]);

    const message = expectError(() =>
        verifyDependencies({
            services,
            uiDependencies: [
                {
                    packageName: "foo",
                    interfaceName: "services.Map2"
                }
            ]
        })
    ).message;
    expect(message).toMatchSnapshot();
});

interface ServiceData {
    package: string;
    name: string;
    requires?: (string | InterfaceSpec)[];
    provides?: (string | InterfaceSpec)[];
}

function mockServices(data: ServiceData[]): ServiceRepr[] {
    return data.map<ServiceRepr>((service) => {
        const name = service.name;
        const packageName = service.package;
        const clazz = class MockService {};
        const dependencies = service.requires?.map<ServiceDependency>((spec, index) => {
            return {
                referenceName: `dep_${index}`,
                ...toInterfaceSpec(spec)
            };
        });
        const interfaces = service.provides?.map(toInterfaceSpec);
        return new ServiceRepr({
            name,
            packageName,
            clazz,
            dependencies,
            interfaces
        });
    });
}

function toInterfaceSpec(spec: string | InterfaceSpec): InterfaceSpec {
    if (typeof spec === "string") {
        return { interfaceName: spec };
    }
    return spec;
}

function getService(
    serviceLookup: ReadonlyServiceLookup,
    interfaceName: string,
    qualifier?: string | undefined
) {
    const result = serviceLookup.lookup({ interfaceName, qualifier });
    if (result.type !== "found") {
        throw new Error(`Failed to find interface ${interfaceName} (qualifier: ${qualifier}).`);
    }
    return result.service;
}
