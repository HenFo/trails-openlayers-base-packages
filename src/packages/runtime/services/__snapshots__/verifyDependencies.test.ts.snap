// Vitest Snapshot v1

exports[`throws for ambiguous service reference 1`] = `"runtime:ambiguous-dependency: Service 'map-user::MapUser' requires the ambiguous interface 'services.Map' as dependency 'dep_0'. Possible choices are map::Map1, map::Map2."`;

exports[`throws when a component depends on itself via a larger cycle 1`] = `"runtime:dependency-cycle: Detected dependency cycle: 'A::a' => 'B::b' ('dep_0' providing 'B') => 'C::c' ('dep_1' providing 'C') => 'A::a' ('dep_2' providing 'A')."`;

exports[`throws when a component directly depends on itself 1`] = `"runtime:dependency-cycle: Detected dependency cycle: 'map::Map' => 'map::Map' ('dep_0' providing 'services.Map')."`;

exports[`throws when a service is not implemented 1`] = `"runtime:interface-not-found: Service 'tools::ExampleTool' requires an unimplemented interface 'services.Map' as dependency 'dep_0'."`;

exports[`throws when an interface is implemented multiple times 1`] = `"runtime:duplicate-interface: Cannot register 'map::Map2' as interface 'services.Map'. 'services.Map' is already provided by service 'map::Map1'. If you intend you have multiple implementations in your application, you must use the 'qualifier' attribute."`;

exports[`throws when the ui requires an interface that is not implemented 1`] = `"runtime:interface-not-found: The UI of package 'foo' requires an unimplemented interface 'services.Map2'."`;
